libfwupdplugin/fu-context.c

```diff
@@ -546,6 +557,184 @@ fu_context_load_quirks (FuContext *self, FuQuirksLoadFlags flags, GError **error
 	return TRUE;
 }
 
+/**
+ * fu_context_get_battery_state:
+ * @self: A #FuContext
+ *
+ * Gets if the system is on battery power, e.g. UPS or laptop battery.
+ *
+ * Returns: a #FuContextBatteryState, e.g. %FU_CONTEXT_BATTERY_STATE_DISCHARGING
+ *
+ * Since: 1.6.0
+ **/
+FuContextBatteryState
+fu_context_get_battery_state (FuContext *self)
+{
+	FuContextPrivate *priv = GET_PRIVATE (self);
+	g_return_val_if_fail (FU_IS_CONTEXT (self), FALSE);
+	return priv->battery_state;
+}
+
+static const gchar *
+fu_context_battery_state_to_string (FuContextBatteryState battery_state)
+{
+	if (battery_state == FU_CONTEXT_BATTERY_STATE_UNKNOWN)
+		return "unknown";
+	if (battery_state == FU_CONTEXT_BATTERY_STATE_CHARGING)
+		return "charging";
+	if (battery_state == FU_CONTEXT_BATTERY_STATE_DISCHARGING)
+		return "discharging";
+	if (battery_state == FU_CONTEXT_BATTERY_STATE_EMPTY)
+		return "empty";
+	if (battery_state == FU_CONTEXT_BATTERY_STATE_FULLY_CHARGED)
+		return "fully-charged";
+	return NULL;
+}
+
+
+/**
+ * fu_context_set_battery_state:
+ * @self: A #FuContext
+ * @battery_state: a #FuContextBatteryState, e.g. %FU_CONTEXT_BATTERY_STATE_DISCHARGING
+ *
+ * Sets if the system is on battery power, e.g. UPS or laptop battery.
+ *
+ * Since: 1.6.0
+ **/
+void
+fu_context_set_battery_state (FuContext *self, FuContextBatteryState battery_state)
+{
+	FuContextPrivate *priv = GET_PRIVATE (self);
+	g_return_if_fail (FU_IS_CONTEXT (self));
+	if (priv->battery_state == battery_state)
+		return;
+	priv->battery_state = battery_state;
+	g_debug ("battery state now %s",
+		 fu_context_battery_state_to_string (battery_state));
+	g_object_notify (G_OBJECT (self), "battery-state");
+}
+
+/**
+ * fu_context_get_battery_level:
+ * @self: A #FuContext
+ *
+ * Gets the system battery level in percent.
+ *
+ * Returns: percentage value, or `0` for unknown
```

Comment:
I actually wonder if this type and return should be a signed integer.  0 can mean a lot of things:
* Dead battery
* No battery
* One battery is 0 and another is something else?

Reply:
What about if we define a `FU_DEVICE_BATTERY_LEVEL_INVALID` as `G_MAXUINT32` or something?

Reply:
or maybe `101` as another option

Reply:
Yes, definitely should define something for this meaning.  No strong feeling across those 3, you pick.

Reply:
Okay, fixed, let me know what you think :)